From 8fe76120565c4a28b6ebb1ccd371f8bfdd8d6d62 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:18:16 -0400
Subject: [fs] nfs: fix nfs4d readlink truncated packet

Message-id: <1410898849-22126-13-git-send-email-bfields@redhat.com>
Patchwork-id: 95843
O-Subject: [Patch RHEL7 012/165] nfs: fix nfs4d readlink truncated packet
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Avi Kivity <avi@cloudius-systems.com>

XDR requires 4-byte alignment; nfs4d READLINK reply writes out the padding,
but truncates the packet to the padding-less size.

Fix by taking the padding into consideration when truncating the packet.

Symptoms:

 # ll /mnt/
 ls: cannot read symbolic link /mnt/test: Input/output error
 total 4
 -rw-r--r--. 1 root root  0 Jun 14 01:21 123456
 lrwxrwxrwx. 1 root root  6 Jul  2 03:33 test
 drwxr-xr-x. 1 root root  0 Jul  2 23:50 tmp
 drwxr-xr-x. 1 root root 60 Jul  2 23:44 tree

Signed-off-by: Avi Kivity <avi@cloudius-systems.com>
Fixes: 476a7b1f4b2c (nfsd4: don't treat readlink like a zero-copy operation)
Reviewed-by: Kinglong Mee <kinglongmee@gmail.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 69bbd9c7b99974f3a701d4de6ef7010c37182a47
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4xdr.c b/fs/nfsd/nfs4xdr.c
index a1c48b4..56ad3d1 100644
--- a/fs/nfsd/nfs4xdr.c
+++ b/fs/nfsd/nfs4xdr.c
@@ -3274,7 +3274,7 @@ nfsd4_encode_readlink(struct nfsd4_compoundres *resp, __be32 nfserr, struct nfsd
 
  wire_count = htonl(maxcount);
  write_bytes_to_xdr_buf(xdr->buf, length_offset, &wire_count, 4);
- xdr_truncate_encode(xdr, length_offset + 4 + maxcount);
+ xdr_truncate_encode(xdr, length_offset + 4 + ALIGN(maxcount, 4));
  if (maxcount & 3)
   write_bytes_to_xdr_buf(xdr->buf, length_offset + 4 + maxcount,
       &zero, 4 - (maxcount&3));
-- 
1.7.1